main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala

```diff
@@ -0,0 +1,454 @@
+/*
+ * Copyright 2022 Paul Butcher
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.api.lsp.provider
+
+import ca.uwaterloo.flix.api.lsp._
+import ca.uwaterloo.flix.language.ast.{Ast, Symbol, Type, TypeConstructor, TypedAst}
+import ca.uwaterloo.flix.language.fmt.{Audience, FormatScheme, FormatType}
+import ca.uwaterloo.flix.language.phase.Parser.Letters
+import ca.uwaterloo.flix.language.phase.Resolver.DerivableSyms
+import ca.uwaterloo.flix.util.InternalCompilerException
+import org.json4s.JsonAST.JObject
+import org.json4s.JsonDSL._
+import org.parboiled2.CharPredicate
+
+/*
+ * CompletionProvider
+ *
+ * Takes a source file, along with the position of the cursor within that file, and returns a list of CompletionItems.
+ *
+ * https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion
+ *
+ * This list is not displayed to the user as-is, the client always both sorts and filters the list (or at least,
+ * VSCode does). Therefore we always need to provide both filterText (currently copied from label) and sortText.
+ *
+ * To ensure that completions are displayed "most useful" first, we preceed sortText with a number as follows:
+ *
+ * 1: High confidence completions (currently with and instance)
+ * 2: Def and Sig
+ * 5: Var
+ * 8: Snippets
+ * 9: Keywords
```

Comment:
For some reason I would actually have expected keywords to be at the highest level, not the lowest level.

What's the reasoning? Just experience?

Reply:
Ideally we want the highest priority suggestion to be the one that's most likely within a given context. Ideally we'd have suggestions that take language structure into account (so we'd know whether we're at a position where we're expecting a top-level keyword, or a variable name, or...). But we don't have that (and won't until we have a partial parser), so it seems to me that the next-best signal is how specific a suggestion is to the context.

So for example `with` suggestions are only made if a lot of context is in place (e.g. a line that starts with `enum`, followed by at least one space, followed by some characters from `with`). If all of that is true, then we can be very confident indeed that the suggestion is relevant.

Keyword suggestions, by contrast, use almost none of the available context (just the characters immediately under the cursor).

The true measure will be how often the top-level suggestion is the right one, so I'm happy to shift this around if a different ordering would make that true more freqently.

Reply:
Makes sense. Let's just see how it plays out :)